### DISCUSSION TITLE
A possible bug on polarization_module.f90

### DISCUSSION THREAD

**Discussion by astroaki** [dullemond/radmc3d-2.0#61] [General]
Dear Kees and developers,

It's Akimasa Kataoka. Thank you for your efforts on keeping RADMC-3D update. 

I sometimes encounter the following error and could not find the solution. I appreciate your feedback.

- I set a protoplanetary disk with full of 140 µm dust grains, and simulate the polarimetric image at the wavelength of 870 µm to see the effects of self-scattering polarization
- The number of photons for the imaging is as huge as 1e8. 
- The error is with message of ' STOP 445'
- This error happens at line 1256 on the polarization_module.f90, where the eps is out of range of 0<eps<1

I slightly dig into the error.
- The error rarely happens when the number of photons is < 1e7, but sometimes happens when number of photons is >1e8.
- When the error happens,  eps = 1.0000000000000007, which is due to the mu (or cos(theta) ) = 1.0000000000000007
- The phot%n(1,2,3) and nobs(1,2,3) are as follows.

 phot%n(1) and nobs(1) are
   4.9714499105734121E-012
   0.0000000000000000     
 phot%n(2) and nobs(2) are
 -0.45243474606228412     
 -0.45243470931178270     
 phot%n(2) and nobs(2) are
  0.89179751096062132     
  0.89179752960521408     

It seems that when the direction of the photon is almost parallel to the direction toward observer, some numerical error happens to make the mu to be mu>1. 

Best regards,
Akimasa

**Comment by dullemond** [comment]
Dear Akimasa,

Interesting bug... Again due to the fact that with large nr of photon, even the improbable becomes probable...

Would you be willing to test the following fix? 

Replace:

```
  call hunt(mui,nmu,mu,jlo)
```

with

```
  if(mu.ge.1.d0) then
     if(mu.gt.1.0000001d0) stop 445
     mu  = 1.d0
     if(mui(nmu).gt.mui(1)) then
        jlo = nmu-1
     else
        jlo = 1
     endif
  elseif(mu.le.-1.d0) then
     if(mu.lt.-1.0000001d0) stop 445
     mu  = -1.d0
     if(mui(nmu).gt.mui(1)) then
        jlo = 1
     else
        jlo = nmu-1
     endif
  else
     call hunt(mui,nmu,mu,jlo)
  endif
```

It is a clunky way to fix it, and I hope that I did not _introduce_ a bug. But I think it should work and be safe.

I would be grateful if you could test if this (1) fixes your problem and (2) does not accidently introduce a new problem / bug. 

Many thanks in advance!

Best wishes,
Kees

**Reply by astroaki** [reply]
Dear Kees,

Thank you for the prompt reply! I'll replace it as suggested. If my student and I find further errors, I will report it here. 

Best regards,
Akimasa
